### PR TITLE
Pipes - Tamira - Hotel 

### PR DESCRIPTION
**Pull request by tvojnar**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   I had to decide how much functionality to put in my Block class versus my Booking class. In other words, if the method to book a room from a block should be in Booking or Block. In the end I went with a halfway decision where the reserve_from_block method in my Booking class preformed some of the functionality of reserving a room from a block, but it also called a method in Block with made it easier to access the instance variables within an instance of Block to complete the reservation.       |
| Describe a concept that you gained more clarity on as you worked on this assignment. |   How to organize classes and how to make many classes work together in a program. |
| Describe a _nominal test_ that you wrote for this assignment. | An example of a nominal test is checking that when a new instance of Reservation was created it was added to the @all_reservations array.   |
| Describe an _edge case test_ that you wrote for this assignment. |  An example of an edge case that I tested was confirming that a room could be booked on the same day that someone checked out of that room.  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I did pretty well. I really tried to follow that pattern, and it really helped to catch bugs right away and fix them before they became entangled with anything else.  |

### CODE REVIEW

File: lib/booking.rb [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+      @all_blocks = []
+      (1..20).each do |number|
+        @all_rooms << Hotel::Room.new(number)
+      end
+    end # initialize
+
+    def valid_dates(start_date, end_date)
+      if end_date < start_date
+        raise BookingError.new("Your checkout day must be after your checkin date! You entered: checkin day = #{start_date} and checkout date = #{end_date}")
+      elsif end_date == start_date
```
Comment (droberts-sea) [inline lib/booking.rb]: Seems to me like this ought to be a responsibility of the `DateRange` class.

File: lib/booking.rb [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+    def check_date_for_reservations(start_date, end_date)
+      # it will have a method that checks if there is a reservation for the date(s) requested
+      # iterate though @all_reservations for each day requested and if the day is included in the Booking then it will look at the Reservation that included that date and add the room(s) in that reservation to a new array not_availible.
+      days = get_date_array(start_date, end_date)
+      date_reservations = []
+      days.each do |day|
+        @all_reservations.each do |res|
+          if res.date_range.include?(day)
+            if !(date_reservations.include?(res))
```
Comment (droberts-sea) [inline lib/booking.rb:83]: All the places that you create a `DateRange`, it seems like you're only using it as a tool to get an array of dates. This indicates that a method might be a better choice here, one that takes a checkin and checkout date and returns an array of dates.

The other option would be to write an instance method, something like `DateRange#overlap?(other)`, that takes another `DateRange` and checks to see if they overlap. Then this complex bit of logic would be hidden away in `DateRange` instead of sitting here in `Booking`.

As long as you've identified "manage a range of dates" as a responsibility, then it probably makes sense to do the later, and keep that responsibility out of the `Booking` class.

This would also make it much easier to test!

File: lib/booking.rb [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        loop_array << "Date range: #{res.date_range[0]} - #{res.date_range[-1] + 1}"
+        nice_format_reservations << loop_array
+      end
+      return nice_format_reservations
+    end #print_reservations
+
+    def availible_rooms(start_date, end_date)
+      reservations = check_date_for_reservations(start_date, end_date)
+      blocks = check_date_for_block(start_date, end_date)
+      reserved = []
```
Comment (droberts-sea) [inline lib/booking.rb:117]: I like this implementation, and appreciate the commenting. It feels like a lot of code, and it might be possible to streamline it a little, but the general idea (get a list of reservations and blocks that overlap the date range, then subtract those rooms from the list of all rooms) is easy to follow.

File: lib/booking.rb [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    end # return_one_block
+
+    def reserve_from_block(block_id, num_of_rooms)
+      id = block_id.upcase
+      all_block_id = return_all_block_ids
+      if !(all_block_id.include?(id))
+        raise BookingError.new("That booking ID does not exist!")
+      end # if
+      reservation_id = @all_reservations.length + 1
+      block_reservations = return_one_block(id).reserve_room(reservation_id, num_of_rooms)
```
Comment (droberts-sea) [inline lib/booking.rb:239]: I love that you call a method here rather than manipulating the block's data directly! Good example of loose coupling!

File: specs/booking_spec.rb [hunk 5]
```diff
@@ -0,0 +1,428 @@
+require_relative 'spec_helper'
+require 'date'
+require 'pry'
+
+describe "Booking" do
+  before do
+    @booking = Hotel::Booking.new
+    @checkin_day = Date.new(2017, 9, 5)
+    @checkout_day = Date.new(2017, 9, 7)
```
Comment (droberts-sea) [inline specs/booking_spec.rb:9]: Good use of a `before` block and instance variables to DRY up these tests.

File: specs/booking_spec.rb [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+    it "will raise BookingError if more rooms are requested then are availible for that date range" do
+      @booking.make_reservation(@checkin_day, @checkout_day, 5)
+      proc{@booking.make_reservation(@checkin_day, @checkout_day, 18)}.must_raise BookingError
+    end # it "will raise BookingError if more rooms are requested then are availible
+
+    it "will be able to book a room the day that it opens up " do
+       @booking.make_reservation(@checkin_day, @checkout_day, 20)
+      #  @booking.make_reservation(@checkout_day, (@checkout_day + 2), 20).must_be_kind_of Array
+       @booking.make_reservation(@checkout_day, (@checkout_day + 2), 20)[0].res_rooms.length.must_equal 20
+    end # it "will be able to book a room the day that it opens up " do
```
Comment (droberts-sea) [inline specs/booking_spec.rb:138]: These tests are a good start, but there are many more interesting cases around whether a room is available that I would like to see. My entire list includes:
- Not available:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Available:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date

File: specs/booking_spec.rb [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+      @booking.must_respond_to :availible_rooms
+    end # it "can be called on @booking" do
+
+    it "will return an array" do
+      @booking.availible_rooms(@checkin_day, @checkout_day).must_be_kind_of Array
+    end #it "will return an array" do
+
+    it "will include a room that is reserved at a differnt time" do
+      room = @booking.make_reservation(@checkin_day, @checkout_day, @num_of_rooms)[0].res_rooms[0]
+
```
Comment (droberts-sea) [inline specs/booking_spec.rb:218]: I like this test case!

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes - good work
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | some - `Booking` seems to have stolen some of the responsibility of the `DateRange` class.
Classes are loosely coupled | yes - good work!
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | yes
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | some - see inline comment
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | yes
**Additional Feedback** |

Great job overall! For the most part your design seems well-thought-out and deliberate, your specific method code is well-organized and easy to read, and your test coverage is solid. Keep up the hard work!
